### DISCUSSION TITLE
KeyNotFoundException occurs in a transaction when trying to insert to planetscale's DB using EF6 or EFcore

### DISCUSSION THREAD

**Discussion by sh1ch** [planetscale/discussion#390] [Bug Reports] [answered]
As the title says, when I try to insert into planetscale's DB using EF6 or EFcore (+MySQL EFcore) in C#, I get an exception on the transaction.

```cs
MySqlContext.ConnectionString = "Server=ap-northeast.connect.psdb.cloud;Database=enterprise-db;user=**;password=pscale_pw_**;SslMode=VerifyFull;";

using var context = new MySqlContext();
using var a = context.Database.BeginTransaction(); // System.Collections.Generic.KeyNotFoundException
```

Error (EF core):

> **System.Collections.Generic.KeyNotFoundException**: 'The given key '0' was not present in the dictionary.'

Error (EF6):

> **KeyNotFoundException**: The given key '29766' was not present in the dictionary.

I think this code was working fine until about early December 2022, but now it doesn't seem to work at all. oddly enough, if I insert directly with MySqlCommand without going through EF, like ExecuteScalar, it works correctly.

```cs
using (var connection = new MySqlConnection(connectionString))
{
	connection.Open();

	var sql1 = "INSERT INTO users (email, first_name, last_name) VALUES ('aa', 'bb', 'cc');";
	var sql2 = "SELECT LAST_INSERT_ID() FROM users;";

	var command = new MySqlCommand(sql1 + sql2, connection);

	long id1 = (long)command.ExecuteScalar();
	long id2 = command.LastInsertedId;

	Console.WriteLine($"inserted id={id1}={id2}"); // ok
}
```

What is the problem with not being able to use EF?

> I have found that using Pomelo.EntityFrameworkCore.MySql as a provider works correctly, but I am not sure how to make it work for MySQL EntityFrameworkCore.

**Comment by orware** [comment, marked as answer]
Hello @sh1ch,

The short answer for you is that I am unsure what the reason is here why using the .NET Entity Framework options aren't working as expected, particularly if they truly had been working before prior to early December 2022 as you shared.

If possible, we can continue troubleshooting in more detail via https://support.planetscale.com/ if a ticket is submitted for us.

I would be curious to find out if you can [enable some enhanced logging](https://learn.microsoft.com/en-us/ef/core/logging-events-diagnostics/) within Entity Framework itself that might show whether there is a specific query/action occurring on the `using var a = context.Database.BeginTransaction(); // System.Collections.Generic.KeyNotFoundException` line that you mentioned seems to be triggering the exception.

However, the `Pomelo.EntityFrameworkCore.MySql` provider you mentioned also seems like a good option/alternative, especially if it is working for you already. I noticed on its README page that it mentions support for `MySQLConnector` and `EF Core` so I was wondering which of those you were using when you ran your working tests? I mainly ask because the developer for MySQLConnector had submitted some [discussion questions](https://github.com/planetscale/discussion/discussions?discussions_q=bgrainger) as well last month so I know some testing with PlanetScale's databases has occurred with it recently that led to some bug reports internally.

**Reply by sh1ch** [reply]
Hello @orware,

Thanks for your comments. It was helpful. First of all, my discussion comment was confusing/omitted in relation to EntityFramework and DataBase Provider, so I tested it again.
(*I also got a comment asking if I used either Connector or EFcore. It helped me because I was a fuzzy on some of my ideas.)

```cs
internal class Program
{
    static void Main(string[] args)
    {
        var program = new Program();

        program.Run();
    }

    public void Run()
    {
        Sample();
    }

    public void Sample()
    {
        MySqlContext.ConnectionString = "Server=ap-northeast.connect.psdb.cloud;Database=***;user=***;password=pscale_pw_***;SslMode=VerifyFull;";

        using var context = new MySqlContext();
        using var a = context.Database.BeginTransaction(); // test
    } 
}
```

```cs
// sample
[Table("system_history")]
public class SystemHistory
{
    [Column("id")]
    public int Id { get; set; }

    [Column("version")]
    public string Version { get; set; } = "";

    [Column("text")]
    public string Text { get; set; } = "";

    [Column("updated_datetime")]
    public DateTime UpdatedDateTime { get; set; }
}
```

## EF6 (MySql.Data.EntityFramework + MySql.EntityFrameworkCore)

NG:

> The given key '15648' was not present in the dictionary.

```cs
public class MySqlConfiguration : DbConfiguration
{
    public MySqlConfiguration()
    {
        var name = MySqlProviderInvariantName.ProviderName;

        SetDefaultConnectionFactory(new SqlConnectionFactory());
        SetProviderFactory(name, new MySql.Data.MySqlClient.MySqlClientFactory());
        SetProviderServices(name, new MySql.Data.MySqlClient.MySqlProviderServices());
    }
}

[DbConfigurationType(typeof(MySqlConfiguration))]
public class MySqlContext : DbContext
{
    public static string ConnectionString { get; set; } = "";
    public DbSet<SystemHistory> SystemHistory => Set<SystemHistory>();
    public MySqlContext() : base(new MySqlConnection(ConnectionString), true) { }
}

```

## EFcore (MySql.Data.EntityFramework + MySql.EntityFrameworkCore)

NG:

> The given key '0' was not present in the dictionary.

```cs
public class MySqlContext : DbContext
{
    public static string ConnectionString { get; set; } = "";
    public DbSet<SystemHistory> SystemHistory => Set<SystemHistory>();

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        optionsBuilder.UseMySQL(ConnectionString, options => { });
        optionsBuilder.LogTo(msg => System.Diagnostics.Debug.WriteLine(msg));
    }
}
```

## EFcore (Microsoft.EntityFrameworkCore + Pomelo.EntityFrameworkCore.MySql)

OK:

```cs
public class MySqlContext : DbContext
{
    public static string ConnectionString { get; set; } = "";
    public DbSet<SystemHistory> SystemHistory => Set<SystemHistory>();

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        var version = new MySqlServerVersion(new Version(8, 0, 23));

        optionsBuilder.UseMySql(ConnectionString, version, options => { });
        optionsBuilder.LogTo(msg => System.Diagnostics.Debug.WriteLine(msg));
    }
}
```

> I noticed on its README page that it mentions support for MySQLConnector and EF Core so I was wondering which of those you were using when you ran your working tests?

What URL is the README page referring to, and does the above answer your question?

**Reply by orware** [reply]
The README page I was referring to was the one for [Pomelo.EntityFrameworkCore.MySql](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#dependencies). In the Dependencies section it seemed to be indicating that it could be used in multiple ways: with MySQLConnector, EF Core, .NET (Core), etc. in the table within that section.

The additional examples you provided should be helpful in attempting to replicate the issue, but it's unclear if it's something that we will be able to address directly or if it'll be more on the MySql.EntityFrameworkCore side of things in some way, or if it would be something we could address somehow within our Vitess layer.

On the MySQL side, the [Entity Framework page](https://dev.mysql.com/doc/connector-net/en/connector-net-entity-framework.html) also lists some notes that talk about full / partial support depending on the MySQL version that I wonder if those could be other reasons for the issues too, and perhaps `Pomelo.EntityFrameworkCore.MySql` helps out with these in other ways. 

I noticed between your `EFcore (MySql.Data.EntityFramework + MySql.EntityFrameworkCore)` example and the last example `EFcore (Microsoft.EntityFrameworkCore + Pomelo.EntityFrameworkCore.MySql)` that the only line difference appears to be the addition of the line: `var version = new MySqlServerVersion(new Version(8, 0, 23));` 

Overall, I'm not sure if we'll be able to find an immediate solution for `MySql.EntityFrameworkCore`, but everything you've shared above should be helpful if we do find that this is a common issue users are encountering when using that option. I'll see if I can spend some time trying to test this too, but it may take me a little while to give it a try on my end.

I hope that the working version with `Pomelo.EntityFrameworkCore.MySql` could potentially still be useful for your needs!

**Comment by kawpls** [comment]
fail: Microsoft.EntityFrameworkCore.Database.Command[20102]
      Failed executing DbCommand (76ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      ALTER DATABASE CHARACTER SET utf8mb4;
Failed executing DbCommand (76ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
ALTER DATABASE CHARACTER SET utf8mb4;
MySqlConnector.MySqlException (0x80004005): alter database is not supported
   at MySqlConnector.Core.ResultSet.ReadResultSetHeaderAsync(IOBehavior ioBehavior) in /_/src/MySqlConnector/Core/ResultSet.cs:line 44
   at MySqlConnector.MySqlDataReader.ActivateResultSet(CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlDataReader.cs:line 127
   at MySqlConnector.MySqlDataReader.CreateAsync(CommandListPosition commandListPosition, ICommandPayloadCreator payloadCreator, IDictionary`2 cachedProcedures, IMySqlCommand command, CommandBehavior behavior, Activity activity, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlDataReader.cs:line 456
   at MySqlConnector.Core.CommandExecutor.ExecuteReaderAsync(IReadOnlyList`1 commands, ICommandPayloadCreator payloadCreator, CommandBehavior behavior, Activity activity, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/CommandExecutor.cs:line 56
   at MySqlConnector.MySqlCommand.ExecuteNonQueryAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlCommand.cs:line 282
   at MySqlConnector.MySqlCommand.ExecuteNonQuery() in /_/src/MySqlConnector/MySqlCommand.cs:line 101
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteNonQuery(RelationalCommandParameterObject parameterObject)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationCommand.ExecuteNonQuery(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationCommandExecutor.ExecuteNonQuery(IEnumerable`1 migrationCommands, IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.UpdateDatabase(String targetMigration, String connectionString, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabaseImpl(String targetMigration, String connectionString, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabase.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
alter database is not supported

**Comment by kawpls** [comment]
Im getting the above issue while trying to update the database with pomelo and EFcore
